packages/affine/block-surface/src/view/mindmap.ts

```diff
@@ -0,0 +1,301 @@
+
+      collapseButton.responseExtension = [16, 16];
+      collapseButton.hidden = hidden;
+      collapseButton.xywh = buttonBound.serialize();
+      collapseButton.groupId = this.model.id;
+      collapseButton.text = collapsed ? node.children.length.toString() : '';
+    }
+
+    if (!alreadyCreated) {
+      collapseButton.opacity = collapsed ? 1 : 0;
+      collapseButton.id = id;
+
+      this._collapseButtons.set(id, collapseButton);
+      this.surface.addLocalElement(collapseButton);
+
+      const buttonView = this.gfx.view.get(id) as GfxElementModelView;
+      const isOnElementBound = (evt: PointerEventState) => {
+        const [x, y] = this.gfx.viewport.toModelCoord(evt.x, evt.y);
+
+        return buttonView.model.includesPoint(
+          x,
+          y,
+          { useElementBound: true },
+          this.gfx.std.host
+        );
+      };
+      let hoveredButton = false;
+      let hoveredNode = false;
+      const updateButtonOpacity = () => {
+        buttonView.model.opacity = hoveredButton || hoveredNode ? 1 : 0;
+      };
+
+      buttonView.on('pointerenter', () => {
+        const latestNode = this.model.getNode(node.id);
+        if (latestNode && latestNode.children.length > 0) {
+          hoveredButton = true;
+          updateButtonOpacity();
+        }
+      });
+      buttonView.on('pointermove', evt => {
+        const latestNode = this.model.getNode(node.id);
+        if (latestNode && latestNode.children.length > 0) {
+          if (isOnElementBound(evt)) {
+            this.gfx.cursor$.value = 'pointer';
+          } else {
+            this.gfx.cursor$.value = 'default';
+          }
+        }
+      });
+      buttonView.on('pointerleave', () => {
+        const latestNode = this.model.getNode(node.id);
+
+        this.gfx.cursor$.value = 'default';
+
+        if (
+          !latestNode ||
+          latestNode.detail.collapsed ||
+          this._selectedButton === buttonView.model
+        ) {
+          return;
+        }
+
+        hoveredButton = false;
+        updateButtonOpacity();
+      });
+      buttonView.on('click', evt => {
+        const latestNode = this.model.getNode(node.id);
+        const telemetry = this.gfx.std.getOptional(TelemetryProvider);
+
+        if (latestNode && isOnElementBound(evt)) {
+          if (telemetry) {
+            telemetry.track('ExpandedAndCollapsed', {
+              page: 'whiteboard editor',
+              segment: 'mind map',
+              type: latestNode.detail.collapsed ? 'expand' : 'collapse',
+            });
+          }
+
+          this.model.toggleCollapse(latestNode!, { layout: true });
+        }
+      });
+
+      const nodeView = this.gfx.view.get(node.id) as GfxElementModelView;
+
+      nodeView.on('pointerenter', () => {
+        hoveredNode = true;
+        updateButtonOpacity();
+      });
+      nodeView.on('pointerleave', () => {
+        hoveredNode = false;
+        updateButtonOpacity();
+      });
+    }
```

Comment:
Do these listeners need to be disposed?

Reply:
Theoretically, yes, but because the handler is only bound once, and the handler will be cleared when the view is destroyed, there’s no need to do so.